kaggle_environments/envs/lux_ai_2022/luxai2022/env.py

```diff
@@ -0,0 +1,618 @@
+                    # all units collide and break
+                    for u in units:
+                        destroyed_units.add(u)
+                elif len(heavy_entered_pos[pos_hash]) > 0:
+                    # all other units collide and break
+                    surviving_unit = heavy_entered_pos[pos_hash][0]
+                    for u in units:
+                        if u.unit_id != surviving_unit.unit_id:
+                            destroyed_units.add(u)
+                    new_units_map_after_collision[pos_hash].append(surviving_unit)
+                else:
+                        # import ipdb;ipdb.set_trace()
+                    # check for stationary heavy unit there
+                    surviving_unit = None
+                    heavy_stationary_unit = None
+                    for u in units:
+                        if u.unit_type == UnitType.HEAVY:
+                            if heavy_stationary_unit is not None:
+                                heavy_stationary_unit = None
+                                # we found >= 2 heavies stationary in a tile where no heavies are entering.
+                                break
+                            heavy_stationary_unit = u
+
+                    if heavy_stationary_unit is not None:
+                        surviving_unit = heavy_stationary_unit
+                    else:
+                        if len(light_entered_pos[pos_hash]) > 1:
+                            # all units collide
+                            surviving_unit = None
+                        elif len(light_entered_pos[pos_hash]) > 0:
+                            # light crashes into stationary light unit
+                            surviving_unit = light_entered_pos[pos_hash][0]
+                    if surviving_unit is None:
+                        for u in units:
+                            destroyed_units.add(u)
+                    else:
+                        for u in units:
+                            if u.unit_id != surviving_unit.unit_id:
+                                destroyed_units.add(u)
+                        new_units_map_after_collision[pos_hash].append(surviving_unit)
+            self.state.board.units_map = new_units_map_after_collision
+
+            for u in destroyed_units:
+                self.destroy_unit(u)
+
+            # for recharging actions, check if unit has enough power. If not, add action back to the queue
+            
+            for unit, recharge_action in actions_by_type["recharge"]:
+                recharge_action: RechargeAction
+                if unit.power < recharge_action.power:
+                    unit.action_queue.insert(0, recharge_action)
+                    # by default actions with repeat=True will be placed to the back of queue
+                    # remove from back of queue if we re-inserted into the front
+                    if recharge_action.repeat:
+                        unit.action_queue = unit.action_queue[:-1]
```

Comment:
aren't we already doing this in `unit.next_action`?

Reply:
`unit.next_action` automatically appends another action to the end if the repeat bit is true

If the recharge action finds that the unit has not recharged enough, we need to add the recharge action to the front so next turn we try recharging again. If it's on repeat mode, then we will end up duplicate repeat actions since we already appended the recharge action to the end (when really we just want to keep it at the front).